server.py

```diff
@@ -146,6 +143,20 @@ def dashboard():
     contests = Contest.select().order_by(Contest.start_time)
     return bottle.template("dashboard.html", contests=contests)
 
+@app.get("/stats")
+@login_required
+def statistics():
+    sub_history_temp = Submission.select(Contest.code, ContestProblems.question, Submission.time, Submission.is_correct)\
+        .where(Submission.user == Session.get(Session.token == bottle.request.get_cookie("s_id")).user)\
```

Comment:
What I usually do is in the login_required function attach a user object called `me` to the bottle request. Then "how" I obtain the `me` is defined in one place and others can use it like `bottle.request.me`

Reply:
I'm sorry @theSage21 but then I don't exactly understand how to do this.

Reply:
so perhaps this is how I would write the login_required function:

```python
def login_required(api):
    def wrapper(*a, **kw):
        try:
            session = Session.get(Session.token == bottle.request.get_cookie('s_id'))
        except Session.DoesNotExist:
            return abort(403, 'please log in')
        bottle.request.session = session
        return api(*a, **kw)
    return wrapper
```

then in the code:

```python

@app.get('/some/api')
@login_required
def my_api():
    SomeTable.select().join(Something).where(Session.id == bottle.request.session, User.id == bottle.request.session.user)
``

Reply:
this ensures that the logic you want to implement to get the cookie is in one place. For example when writing APIs it is common to have tokens in the `Authorization` header OR the cookie. Then this kind of code becomes useful.

Reply:
Thanks a lot.